nilearn/glm/first_level/first_level.py

```diff
@@ -122,25 +141,39 @@ def run_glm(Y, X, noise_model='ar1', bins=100, n_jobs=1, verbose=0):
     # Create the model
     ols_result = OLSModel(X).fit(Y)
 
-    if noise_model == 'ar1':
+    if 'ar' in noise_model:
+        ar_order = int(re.split('ar', noise_model)[1])
         # compute and discretize the AR1 coefs
-        ar1 = (
-            (ols_result.residuals[1:]
-             * ols_result.residuals[:-1]).sum(axis=0)
-            / (ols_result.residuals ** 2).sum(axis=0)
-        )
+        ar1 = [_yule_walker(ols_result.residuals[:, res].reshape(-1, 1).T,
+                            order=ar_order)
+               for res in range(ols_result.residuals.shape[1])]
+        if len(ar1[0]) == 1:
+            ar1 = np.asarray(ar1).ravel()
         del ols_result
-        ar1 = (ar1 * bins).astype(np.int) * 1. / bins
-        # Fit the AR model acccording to current AR(1) estimates
+        for idx in range(len(ar1)):
+            ar1[idx] = (ar1[idx] * bins).astype(np.int) * 1. / bins
+        # Fit the AR model acccording to current AR(N) estimates
         results = {}
         labels = ar1
+
         # Parallelize by creating a job per ARModel
-        vals = np.unique(ar1)
+        if type(ar1[0]) is np.float64:
+            vals = np.unique(ar1)
+        else:
+            vals = ar1
+
         ar_result = Parallel(n_jobs=n_jobs, verbose=verbose)(
-            delayed(_ar_model_fit)(X, val, Y[:, labels == val])
+            delayed(_ar_model_fit)(X, val, Y[:, [np.all(lab == val)
+                                                 for lab in labels]])
             for val in vals)
-        for val, result in zip(vals, ar_result):
-            results[val] = result
+
+        if type(ar1[0]) is np.float64:
+            for val, result in zip(vals, ar_result):
+                results[val] = result
+        else:
+            labels = ['_'.join([str(v) for v in val]) for val in vals]
```

Comment:
I wasn't sure what the labels should be for AR(N) models. So here I am just joining the values with an underscore. But if I understand better what the clustering is used (comment above) then maybe this will answer itself.